app/models/assignment.rb

```diff
@@ -13,23 +13,38 @@ class Assignment < ActiveRecord::Base
   validates :mixpanel_result, inclusion: { in: %w(success failure) }, allow_nil: true
   validate :variant_must_exist
 
-  scope :unsynced_to_mixpanel, -> { where("mixpanel_result = 'failure' OR mixpanel_result IS NULL") }
-
   normalize_attributes :mixpanel_result
 
   delegate :name, to: :split, prefix: true
 
-  class << self
-    def to_hash
-      Hash[all.includes(:split).map { |a| [a.split.name.to_sym, a.variant.to_sym] }]
-    end
+  scope :for_presentation, ->(app_build: nil) do
+    q = excluding_decision_overrides
+    app_build.present? ? q.for_app_build(app_build) : q.for_active_splits
+  end
 
-    def for_presentation(built_at: nil)
-      q = joins(:split)
-        .where("splits.decided_at is null or assignments.updated_at > splits.decided_at")
-      q = built_at.nil? ? q.where("splits.finished_at is null") : q.where("splits.finished_at > ?", built_at)
-      q
-    end
+  scope :for_app_build, ->(app_build) do
+    for_active_splits(as_of: app_build.built_at)
+      .excluding_incomplete_features(app_id: app_build.app_id, version: app_build.version)
+  end
+
+  scope :excluding_decision_overrides, -> do
+    joins(:split).where('splits.decided_at is null or assignments.updated_at > splits.decided_at')
+  end
+
+  scope :for_active_splits, ->(as_of: nil) do
+    joins(:split).merge(Split.active(as_of: as_of))
+  end
+
+  scope :excluding_incomplete_features, ->(app_id:, version:) do
+    joins(:split).where(<<~SQL, app_id, AppVersion.new(version).to_pg_array)
+      splits.feature_gate = false
+      or exists (
+        select 1 from feature_completions fc
+        where fc.app_id = ?
+          and fc.split_id = assignments.split_id
+          and fc.version <= ?
```

Comment:
So, I get that this is where the very elegant magic happens -- and ultimately why versions are treated as value objects (so that they can cast themselves to and from pg arrays).

I'm curious about how/whether this can be indexed. (Trying to read the pg docs, I can't tell if a B-tree index explicitly doesn't work for array columns, or if it works but not for every array operation.)

The alternative, in my mind, would be to store three separate integers, which can most certainly be indexed but would, I imagine, result in a more complex query.

Reply:
I’m not using version to restrict my working set that significantly, so indexing is probably not worth it - the set of assignments I have as a visitor is the highest cardinality index.

Reply:
Makes sense 👍